Backpack/src/main/java/net/skyscanner/backpack/dialog/BpkDialog.kt

```diff
@@ -0,0 +1,120 @@
+import androidx.annotation.DrawableRes
+import androidx.core.content.res.ResourcesCompat
+import net.skyscanner.backpack.button.BpkButton
+
+open class BpkDialog(
+  context: Context,
+  val style: Style = Style.ALERT
+) : Dialog(context, 0) {
+
+  data class Icon(@DrawableRes val drawableRes: Int, @ColorRes val colorRes: Int)
+
+  enum class Style {
+    ALERT, BOTTOM_SHEET
+  }
+
+  private var viewHolder: ViewHolder? = null
+  private val variables = DialogVariables(context)
+
+  init {
+    setupDialog()
+  }
+
+  var title = ""
+    set(value) {
+      field = value
+      viewHolder?.title?.text = value
+    }
+
+  var description = ""
+    set(value) {
+      field = value
+      viewHolder?.description?.text = value
+    }
+
+  var icon: Icon? = null
+    set(value) {
+      field = value
+      viewHolder?.iconView?.icon = icon
+    }
+
+  fun addActionButton(button: BpkButton) {
+    button.layoutParams = LinearLayout.LayoutParams(
+      LinearLayout.LayoutParams.MATCH_PARENT,
+      LinearLayout.LayoutParams.WRAP_CONTENT
+    ).apply {
+      setMargins(0, 0, 0, variables.buttonStackItemSpace)
+    }
+    viewHolder?.contentLayout?.addView(button)
+  }
+
+  private fun setupDialog() {
+    if (style == Style.BOTTOM_SHEET) {
+      window?.setGravity(Gravity.BOTTOM)
+      window?.setWindowAnimations(R.style.Bpk_dialog_animation)
+    }
+
+    requestWindowFeature(Window.FEATURE_NO_TITLE)
+    setContentView(setUpContentView())
+    window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))
+
+    val metrics = DisplayMetrics()
+    val windowManager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager
+    windowManager.defaultDisplay.getMetrics(metrics)
+    window?.setLayout((metrics.widthPixels * 0.85).toInt(), RelativeLayout.LayoutParams.WRAP_CONTENT)
+  }
+
+  private fun setUpContentView(): View {
+    val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
+    val view = inflater.inflate(R.layout.bpk_dialog, null)
+
+    viewHolder = ViewHolder(view).apply {
+      val params = contentLayout.layoutParams as RelativeLayout.LayoutParams
+      params.topMargin = variables.contentContainerMarginTop
+      contentLayout.background = getContentBackground()
+      contentLayout.setPadding(
+        variables.contentContainerPaddingHorizontal,
+        variables.contentContainerPaddingTop,
+        variables.contentContainerPaddingHorizontal,
+        variables.contentContainerPaddingBottom
+      )
+    }
+
+    return view
+  }
+
+  private fun getContentBackground(): GradientDrawable {
+    val shape = GradientDrawable()
+    shape.shape = GradientDrawable.RECTANGLE
+    val radii = variables.contentRadius
+    shape.cornerRadii = floatArrayOf(radii, radii, radii, radii, radii, radii, radii, radii)
+    shape.setColor(ResourcesCompat.getColor(context.resources, R.color.bpkWhite, context.theme))
+    return shape
+  }
+
+  private class ViewHolder(container: View) {
```

Comment:
We should consider moving to kotlinx to avoid viewholders

Reply:
How would that work?